File: app/src/main/java/se/accidis/fmfg/app/services/MaterialsRepository.java
package se.accidis.fmfg.app.services;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import org.json.JSONArray;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import se.accidis.fmfg.app.model.Material;
import se.accidis.fmfg.app.utils.IOUtils;

/**
 * Repository for the materials list.
 */
public final class MaterialsRepository {
    private static final String ADR_JSON_ASSET = "ADR.json";
    private static final String TAG = MaterialsRepository.class.getSimpleName();
    private final Context mContext;
    private final Set<String> mFavoriteMaterials;
    private final Preferences mPrefs;
    private static MaterialsRepository mInstance;
    private List<Material> mList;
    private OnLoadedListener mOnLoadedListener;

    private MaterialsRepository(Context context) {
        mContext = context.getApplicationContext();
        mPrefs = new Preferences(mContext);
        mFavoriteMaterials = mPrefs.getFavoriteMaterials();
    }

    public static MaterialsRepository getInstance(Context context) {
        return (null == mInstance ? (mInstance = new MaterialsRepository(context)) : mInstance);
    }

    public void addFavoriteMaterial(Material material) {
        String key = material.toUniqueKey();
        mFavoriteMaterials.add(key);
        mPrefs.setFavoriteMaterials(mFavoriteMaterials);
    }

    public void beginLoad() {
        if (null != mList) {
            Log.d(TAG, "Assets already loaded, nothing to do.");
            if (null != mOnLoadedListener) {
                mOnLoadedListener.onLoaded(mList);
            }

            return;
        }

        Log.d(TAG, "Loading assets.");
        LoadTask loadTask = new LoadTask();
        loadTask.execute();
    }

    public boolean isFavoriteMaterial(Material material) {
        String key = material.toUniqueKey();
        return mFavoriteMaterials.contains(key);
    }

    public void removeFavoriteMaterial(Material material) {
        String key = material.toUniqueKey();
        mFavoriteMaterials.remove(key);
        mPrefs.setFavoriteMaterials(mFavoriteMaterials);
    }

    public void setOnLoadedListener(OnLoadedListener listener) {
        mOnLoadedListener = listener;
    }

    public interface OnLoadedListener {
        void onException(Exception ex);

        void onLoaded(List<Material> list);
    }

    private final class LoadTask extends AsyncTask<Void, Void, Void> {
        @Override
        protected Void doInBackground(Void... params) {
            try {
                String str = readAssetsFile();
                JSONArray jsonArray = new JSONArray(str);

                ArrayList<Material> list = new ArrayList<>(jsonArray.length());
                for (int i = 0; i < jsonArray.length(); i++) {
                    list.add(Material.fromJSON(jsonArray.getJSONObject(i)));
                }

                mList = list;
                Log.i(TAG, "Finished loading assets (" + mList.size() + " items loaded).");

            } catch (Exception ex) {
                Log.e(TAG, "Failed to load assets.", ex);

                if (null == mList && null != mOnLoadedListener) {
                    mOnLoadedListener.onException(ex);
                }
            }

            if (null != mList && null != mOnLoadedListener) {
                mOnLoadedListener.onLoaded(mList);
            }

            return null;
        }

        private String readAssetsFile() throws IOException {
            return IOUtils.readToEnd(mContext.getAssets().open(ADR_JSON_ASSET));
        }
    }
}

File: app/src/main/java/se/accidis/fmfg/app/services/MaterialsRepository.kt
package se.accidis.fmfg.app.services

import android.content.Context
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import se.accidis.fmfg.app.model.Material
import se.accidis.fmfg.app.model.MaterialSource
import se.accidis.fmfg.app.utils.Resource
import timber.log.Timber

/**
 * Repository for the materials list.
 */
class MaterialsRepository private constructor(context: Context) {
    private val context: Context = context.applicationContext
    private val _materialsResource = MutableStateFlow<Resource<List<Material>>>(Resource.Loading)
    val materialsResource: StateFlow<Resource<List<Material>>> = _materialsResource.asStateFlow()

    private val repositoryScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    fun beginLoad() {
        val current = _materialsResource.value
        if (current is Resource.Success && current.data.isNotEmpty()) {
            Timber.d("Assets already loaded, nothing to do.")
            return
        }

        Timber.d("Loading assets.")
        _materialsResource.value = Resource.Loading
        repositoryScope.launch {
            try {
                val loadedMaterials = withContext(Dispatchers.IO) {
                    val adrJob = async { loadFromAsset(ADR_S_JSON_ASSET, MaterialSource.ADR_S) }
                    val amkatJob = async { loadFromAsset(AMKAT_JSON_ASSET, MaterialSource.AMKAT) }
                    adrJob.await() + amkatJob.await()
                }

                _materialsResource.value = Resource.Success(loadedMaterials)
                Timber.i("Finished loading assets (%d items loaded).", loadedMaterials.size)
            } catch (ex: Exception) {
                Timber.e(ex, "Failed to load assets.")
                _materialsResource.value = Resource.Error(ex)
            }
        }
    }

    private fun loadFromAsset(assetName: String, source: MaterialSource): List<Material> {
        val str = context.assets.open(assetName).bufferedReader().use { it.readText() }
        val jsonArray = JSONArray(str)
        val list = ArrayList<Material>(jsonArray.length())
        for (i in 0 until jsonArray.length()) {
            list.add(Material.fromJSON(jsonArray.getJSONObject(i), source))
        }
        return list
    }

    companion object {
        private const val ADR_S_JSON_ASSET = "adr-s.mini.json"
        private const val AMKAT_JSON_ASSET = "amkat.mini.json"
        private var singleton: MaterialsRepository? = null

        @JvmStatic
        fun getInstance(context: Context): MaterialsRepository =
            singleton ?: synchronized(this) {
                singleton ?: MaterialsRepository(context).also { singleton = it }
            }
    }
}
